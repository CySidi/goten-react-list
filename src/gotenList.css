@font-face {
  font-family: 'Glyphicons Halflings';
  src: url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.eot');
  src: url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.eot?#iefix') format('embedded-opentype'),
       url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.woff2') format('woff2'),
       url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.woff') format('woff'),
       url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.ttf') format('truetype'),
       url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.svg#glyphicons_halflingsregular') format('svg');
}

.actions {
    float: right;
}

.actionTitle {
    padding-left: 5px;
}

.title {
    padding-right: 5px;
}

.button-icon {
    border: 0px; 
    background-color: rgba(0, 0, 0, 0);
}

.glyphicon {
  position: relative;
  top: 1px;
  display: inline-block;
  font: normal normal 16px/1 'Glyphicons Halflings';
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  margin-right: 4px;
}

.glyphicon-pencil:before{content:"\270f";}
.glyphicon-trash:before{content:"\e020";}
.glyphicon-search:before{content:"\e003";}