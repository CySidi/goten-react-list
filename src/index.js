import React, { Component } from 'react'

import Table from 'react-bootstrap/lib/Table'
import Glyphicon from 'react-bootstrap/lib/Glyphicon'
import Button from 'react-bootstrap/lib/Button'

import './gotenList.css'


const bootstrapLink = <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u"
    crossOrigin="anonymous" />

const GotenListState = {
    items: []
}

export class GotenList extends Component {
    constructor(props){
        super(props)
        this.state = GotenListState
        this.keyItem = 0
    }

    removeItems() {
        this.setState({items:[]})
    }

    addItem(item, actions) {
        this.setState(prevState => ({
            items: [...prevState.items, {
                item: item,
                key: this.keyItem++,
                ...actions
            }]
        }))
    }

    addItemIteratively(item, numberOfItems=1, actions) {
        for (let index = 0; index < numberOfItems; index++) {
            this.addItem(item, actions)
        }
    }
    
    render(){
        return (
            <div>
                {bootstrapLink}
                <Table striped condensed hover>
                    <thead>
                        <tr>
                            <th>{this.props.title}</th>
                            <th>
                                <label className='actions'>
                                    {this.props.actionsTitle}
                                </label>
                            </th>
                        </tr>
                    </thead>
                    <tbody>
                        {this._getItems()}
                    </tbody>
                </Table>
            </div>
        )
    }
    
    _removeItem = async (item) => {
        if (this.props.onRemove)
            this.props.onRemove()
        else if (item.onRemove)
            item.onRemove()
        this.setState(prevState => ({
            items: [...prevState.items.filter(prevItem => prevItem.key !== item.key)]
        }))
    }

    _getOneAction(logic, name) {
        return (
            <th>
                { logic &&
                <Button
                    bsClass='button-icon'
                    onClick={logic}
                >
                    <Glyphicon glyph={name} />
                </Button>
                }
            </th>
        )
    }

    _getActions(item) {
        return (
            <table className='actions'>
                <tbody>
                    <tr>
                        {item.onSearch ?
                            this._getOneAction(item.onSearch, 'search')
                            : this._getOneAction(this.props.onSearch, 'search')}
                        {item.onEdit ?
                            this._getOneAction(item.onEdit, 'pencil')
                            : this._getOneAction(this.props.onEdit, 'pencil')}
                        {this._getOneAction(_ => this._removeItem(item), 'trash')}
                    </tr>
                </tbody>
            </table>
        )
    }

    _getItems() {
        return this.state.items.map( item => {
            return (
                <tr key={item.key}>
                    <th>
                        {item.item}
                    </th>
                    <th>
                        {this._getActions(item)}
                    </th>
                </tr>
            )
        })
    }
}
