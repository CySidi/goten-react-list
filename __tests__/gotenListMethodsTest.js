import React from 'react'
import renderer from 'react-test-renderer'

import { GotenList } from '../src'


describe('GotenList methods tests', () => {

    it('check addItem', () => {
        const component = renderer.create(
            <GotenList />
        )
        component.getInstance().addItem(<item>Component</item>)
        expect(component.root.findAllByType('item')).toHaveLength(1)
    })

    it('check add five items with addItem', () => {
        const items = 5
        const component = renderer.create(
            <GotenList />
        )
        for (let index = 0; index < items; index++) {
            component.getInstance().addItem(<item>Component</item>)
        }
        expect(component.root.findAllByType('item')).toHaveLength(items)
    })


    it('check add five items with addItem and show its with mergeColumns', () => {
        const items = 5
        const component = renderer.create(
            <GotenList 
                mergeColumns={true}
            />
        )
        for (let index = 0; index < items; index++) {
            component.getInstance().addItem(<item>Component</item>)
        }
        expect(component.root.findAllByType('item')).toHaveLength(items)
        component.toJSON()
        expect(component).toMatchSnapshot()
    })

    it('check add five items with addItemIteratively', () => {
        const items = 5
        const component = renderer.create(
            <GotenList />
        )
        component.getInstance().addItemIteratively(<item>Component</item>, items)
        expect(component.root.findAllByType('item')).toHaveLength(items)
    })

    it('check add one items with addItemIteratively', () => {
        const items = 1
        const component = renderer.create(
            <GotenList />
        )
        component.getInstance().addItemIteratively(<item>Component</item>, items)
        expect(component.root.findAllByType('item')).toHaveLength(items)
    })

    it('check addItem with an action', () => {
        const component = renderer.create(
            <GotenList />
        )
        component.getInstance().addItem(<item>Component</item>, {onEdit: _ => null})
        component.toJSON()
        expect(component).toMatchSnapshot()
    })

    it('check add five items with addItemIteratively and with an action', () => {
        const items = 5
        const component = renderer.create(
            <GotenList />
        )
        component.getInstance().addItemIteratively(<item>Component</item>, items,
            { onEdit: _ => null })
        component.toJSON()
        expect(component).toMatchSnapshot()
    })

    it('check remove all items', () => {
        const items = 5
        const component = renderer.create(
            <GotenList />
        )
        component.getInstance().addItemIteratively(<item>Component</item>, items,
            { onEdit: _ => null })
        expect(component.root.findAllByType('item')).toHaveLength(items)
        component.getInstance().removeItems()
        expect(component.root.findAllByType('item')).toHaveLength(0)
    })

    it('check addItem with an array of components', () => {
        const component = renderer.create(
            <GotenList />
        )
        component.getInstance().addItem([<item>Component1</item>, <item>Component2</item>])
        expect(component.root.findAllByType('item')).toHaveLength(2)
    })
})
