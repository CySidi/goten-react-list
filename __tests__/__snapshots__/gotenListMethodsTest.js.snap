// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GotenList methods tests check add five items with addItem and show its with mergeColumns 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th />
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  >
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      />
    </tr>
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      />
    </tr>
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      />
    </tr>
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      />
    </tr>
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      />
    </tr>
  </tbody>
</table>
`;

exports[`GotenList methods tests check add five items with addItemIteratively and with an action 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th />
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  >
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      >
        <button
          className="button-icon button-icon-default"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          type="button"
        >
          <span
            className="glyphicon glyphicon-pencil"
          />
        </button>
      </td>
    </tr>
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      >
        <button
          className="button-icon button-icon-default"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          type="button"
        >
          <span
            className="glyphicon glyphicon-pencil"
          />
        </button>
      </td>
    </tr>
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      >
        <button
          className="button-icon button-icon-default"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          type="button"
        >
          <span
            className="glyphicon glyphicon-pencil"
          />
        </button>
      </td>
    </tr>
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      >
        <button
          className="button-icon button-icon-default"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          type="button"
        >
          <span
            className="glyphicon glyphicon-pencil"
          />
        </button>
      </td>
    </tr>
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      >
        <button
          className="button-icon button-icon-default"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          type="button"
        >
          <span
            className="glyphicon glyphicon-pencil"
          />
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`GotenList methods tests check addItem with a Row component 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th>
        column1
      </th>
      <th>
        column2
      </th>
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  >
    <tr>
      <td>
        text from column1
      </td>
      <td>
        text from column2
      </td>
      <td
        align="right"
      />
    </tr>
  </tbody>
</table>
`;

exports[`GotenList methods tests check addItem with an action 1`] = `
<table
  className="table table-striped table-condensed table-hover"
  width="100%"
>
  <thead
    align="left"
    className="table-thead"
  >
    <tr>
      <th />
      <th
        align="right"
      />
    </tr>
  </thead>
  <tbody
    align="left"
  >
    <tr>
      <td>
        <item>
          Component
        </item>
      </td>
      <td
        align="right"
      >
        <button
          className="button-icon button-icon-default"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          type="button"
        >
          <span
            className="glyphicon glyphicon-pencil"
          />
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
